### ISSUE TITLE
Beta Launch Pre-Flight Check

### ISSUE THREAD

**Issue by docuracy** [WorldHistoricalGazetteer/whg3#214]
- [x] Slow load (see comment below)
- [x] ~_Perhaps_, remove `carouselMetadata` from Home Page context/HTML and fetch with AJAX.~
- [x] Throttle and limit calls to `/task_progress/` in `tileset-tools.js`.
- [x] Ecolayers in WHG basemap need to be invisible, ~and toggled on & off during Portal initialisation to get values.~ Fix: proved impossible as proposed so created additional base style for Portal with ecolayers; default no longer has them.
- [x] Portal Page elevation data for Tblisi looks wrong (0 metres).
- [x] Portal Page height too large to accommodate footer without scroll on "standard" screen size?
- [x] https://dev.whgazetteer.org/places/13970069/portal/ includes a `null` entry among the "types" in the `.gloss`.
- [x] https://dev.whgazetteer.org/tileset_management/ : needs_tileset() got an unexpected keyword argument 'ignore_result'
- [x] http://localhost:8001/datasets/54/places : click on rows - zoom way off-centre - tileset points seem to be over-approximated. **Fix: force tippecanoe tileset generation up to -z14 instead of -zg (guess)**
- [x] https://dev.whgazetteer.org/datasets/1209/places: Place Detail and Map should both have wait spinners.
- [x] Now implementing by default the basemap which includes OSM layers. Which of these should **NOT** have that, @kgeographer? **comment below**
    - [x] Home http://localhost:8001/
    - [x] Gallery http://localhost:8001/datasets/gallery/datasets/
    - [ ] Search http://localhost:8001/search/
    - [ ] Portal http://localhost:8001/places/portal/ (has already)
    - [ ] Collection Places http://localhost:8001/collections/11/browse_pl
    - [ ] Dataset Places http://localhost:8001/datasets/12/places
    - [ ] Dataset Browse http://localhost:8001/datasets/12/browse
    - [ ] Areas http://localhost:8001/areas/create/
    - [ ] Tasks http://localhost:8001/datasets/7/addtask _<-- Accessed through "My Data", then pick a "reconciling" dataset_
    - [ ] Review http://localhost:8001/datasets/9/review/... _<-- Accessed through "My Data", then pick a "reconciling" dataset_
- [x] Clustering in `/app/utils/cluster_geometries.py` is prone to causing Out of Memory errors. Switched from Agglomerative to KMeans clustering: seems quicker locally, but **needs to be tested with large datasets on dev server**.
- [x] set home page to `v30a4`
- [x] Convert any timespan years to integers in `portal.js`
- [x] top menu cursor links inaccessible; hotfixed commenting of display:block. Also permanently fixed in `whg-modal.css`
- [x] ~Google Analytics code is getting HTTP 204 errors - is it obsolete?~ No, this is normal - see [here](https://stackoverflow.com/questions/71752244/ga4-event-validation-only-returns-204).
- [x] https://dev.whgazetteer.org/: No typeahead, and **Search not working at all!** 
`home.bundle.js:985 Uncaught ReferenceError: csrfToken is not defined at initiateSearchHome (home.bundle.js:985:20)` added `window.csrfToken = "{{ csrf_token }}";`
- [x] ~https://dev.whgazetteer.org/dashboard_admin/: Dataset list loading spinner needed.~ Deferred: probably only a temporary problem.
- [x] ~https://dev.whgazetteer.org/dashboard_admin/: Table sorting is very slow, and spinner is mis-located. `https://dev.whgazetteer.org/main/datasets_list` is not paginated, is it, so can't the table be sorted in the browser without an additional call to the server? **kg: this is quite a convoluted setup, loading '_list' pages into the dashboard_admin page. Tricky because each tab has different search/filter requirements. I think doing something else is a major refactor. This page is seen only by the team, so a spinner should be fine for now; low priority** pls defer~
- [x] https://dev.whgazetteer.org/datasets/gallery/datasets/: Carousel horses include datasets with `null` images - do not include these in the `.ds-card-blurb` divs.
- [x] https://dev.whgazetteer.org/datasets/gallery/datasets/: Carousel horse titles should be linked to the associated Dataset/Collection browse page.
- [x] https://dev.whgazetteer.org/search/: search failure should provide feedback in the UI (e.g. calls to `search/suggestions` which return an error).

**Comment by docuracy** [comment]
@kgeographer 

Some portal pages (well, at least one - I can't check others because the server is rebooting, I think) are failing to load due to malformed timespans, which come directly from the `timespans` JSONField in the `Place` model, for example (https://dev.whgazetteer.org/places/13970069/portal/):

`... [200, 1933], [500, 1800], &#x27;[&#x27;, &#x27;2&#x27;, &#x27;0&#x27; ...`

It seems that integer values might have been stored as strings within encoded quote marks (or possibly the square brackets and commas are enclosed in encoded quote marks, though this seems less likely).

**Comment by kgeographer** [comment]
I see what the issue is with timespans, and have partially fixed it - some values were malformed in the migration, enclosing the entire value in double quotes; got rid of those. Also found that in both v2 and now v3 some timespans have quoted values within the span, like ["830", "930"] or ["830", 1200]. Apart from the cause, what must a timespans value look like for your code to parse it properly, because I can normalize them in sql. Should they all be nested arrays of integers?

**Comment by docuracy** [comment]
KG: slow load on server, not locally so it's the data. removed collections from `featured` status, no help. the browser network console shows 'pending' [figs below]. best to wait & see whether adding collection data solves it. 

![image](https://github.com/WorldHistoricalGazetteer/whg3/assets/1753945/fd94e1ea-0751-4611-97fa-c0d550aa2859)
![image](https://github.com/WorldHistoricalGazetteer/whg3/assets/1753945/5cd0e5a5-f8af-42c0-a75a-4cf75d2ac880)

**Comment by kgeographer** [comment]
I fixed the timespans thing - that is why I checked it off

**Comment by docuracy** [comment]
>- [x] https://dev.whgazetteer.org/: No typeahead, and **Search not working at all!** 
>`home.bundle.js:985 Uncaught ReferenceError: csrfToken is not defined at initiateSearchHome (home.bundle.js:985:20)` added >`window.csrfToken = "{{ csrf_token }}";`

@kgeographer : the csrfToken is loaded in base_webpack, but I had failed to make it a global (window) variable. It's now fixed there (and deployed throughout the webpack codebase) so the `window.csrfToken = "{{ csrf_token }}";` can be removed from the home templates.

**Comment by docuracy** [comment]
Added an `error-modal.js` module which can be used anywhere in the codebase to display a modal dialog like this one. The `title` (default "ERROR"), `error`, and `timeout` (in seconds, default: do not close) parameters are optional.

'''
function errorModal(message, title, error, timeout)
'''

![Image](https://github.com/WorldHistoricalGazetteer/whg3/assets/42514781/6c8650dd-3b87-468b-ba4a-6d9bd310dada)

**Comment by kgeographer** [comment]
re: basemaps, only the homw and gallery pages must have 'bare topography' - areas and tasks can also, no need for context

on any that offer a zoom to feature, there has to be a buffer - most zoom too far in by far

**Comment by docuracy** [comment]
> re: basemaps, only the home and gallery pages must have 'bare topography' - areas and tasks can also, no need for context in those - well maybe countries would be nice to see
> 
> on any that offer a zoom to feature, there has to be a buffer - most zoom too far in by far

There are now 3 different map styles in use:
- **WHG Basic Light** (on Home & Gallery pages)
- **WHG-Portal** which has the full works (Portal page only)
- **WHG** which has the full works except ecolayers (everywhere else)

If it becomes apparent that anything is annoying on the areas and tasks pages, it's easy enough to define an additional style. And next time I'll be sure to avoid spurious trailing commas in JSON lists (which are legal in Python data structures): the tileserver doesn't do dodgy JSON.

There's now a global `defaultZoom` variable set to the value `8`  in `base.js`, which is applied as a maximum zoom parameter wherever the `whg_maplibre.fitViewPort` method is called - except on the home page and in carousels. It seems to me to be about right in most contexts, but can now easily be adjusted.

**Comment by docuracy** [comment]
The codebase _was_ using a UMD version of `spin.js`, which according to the [documentation](https://spin.js.org/) "is only available as a temporary workaround".

It is now (with the next PR) using the recommended ECMAScript module, bundled with Webpack and extended in `base.js` to allow initialisation and stopping with simple JQuery selectors. Scaling is determined automatically based on the size of the target container.

Examples:

```
$('#drftable_list, #detail').spin(); // Start spinners on single or multiple elements - can also use `.class` selectors

$('#detail, .maplibregl-control-container').stopSpin(); // Stop existing spinners on single or multiple elements
```

**Comment by docuracy** [comment]
## https://dev.whgazetteer.org/ Testing

### UI Issues

#### Portal Page
- [ ] Map in `.no-results` should hide Drawing and Temporal filters and Fullscreen control.

#### Search Page
- [ ] Results ought to include timespans, which can currently be deduced only by moving the time slider

### Data Issues (probably)

- [ ] https://dev.whgazetteer.org/places/13120162/portal/ Blank - search for "London" claims 2 linked records
- [ ] http://localhost:8001/places/12346695/portal/ Blank - search for "Grenada" claims 2 linked records
- [ ] https://dev.whgazetteer.org/places/5907944/detail
    - Linked Places record link HTTP 404: https://dev.whgazetteer.org/api/db/?id=5907944
    -

**Comment by docuracy** [comment]
![Screenshot_20240531-165152.png](https://github.com/WorldHistoricalGazetteer/whg3/assets/42514781/885746f4-27f9-479d-83a0-f7ba30f51987)

Got this when trying to test the Contact Form on dev using mobile phone. Need to check that csrfToken is properly reaching the modal - it _is_ present on the page, allowing search POST requests.

**Comment by docuracy** [comment]
> KG: slow load on server, not locally so it's the data. removed collections from `featured` status, no help. the browser network console shows 'pending' [figs below]. best to wait & see whether adding collection data solves it.
> 

Removing the Roman Quarry dataset's featured flag reduces Home Page load time from 6 seconds to 2 seconds on `dev`, but makes no discernable difference locally. It was the only featured dataset that qualified for heatmapping (500 or more places all having Point or MultiPoint geometry).

Our annoying friend has ideas as to why there should be such a large time-difference between running the `display_mode` @property function locally and on the dev server, including for example: 

```
CREATE INDEX idx_place_geom_place ON PlaceGeom(place);
```

Do the databases have the same indexes?

That friend has also suggested more efficient code for calculating the `display_mode`. I propose that we implement that and use it to add a `points_only` field to the database tables for both Datasets and Collections the first time the `display_mode` is called (as we did for `coordinate_density`), so that carousel_data can be loaded in a jiffy.

**Comment by kgeographer** [comment]
> We might need something like:
> 
> CREATE INDEX idx_place_geom_place ON PlaceGeom(place);

**added that. home page loads fast, including with Stone Quarries** 

I had removed indexes on dev because they made unique testing during migration impossibly slow, then recreated the spatial gist one but not the place_id one <= which is the FK...doh

**Comment by docuracy** [comment]
The server response for `datasets.utils.fetch_mapdata_ds` is slow:

![Image](https://github.com/WorldHistoricalGazetteer/whg3/assets/42514781/6c89d526-0aba-4a9c-870c-a61806695d22)

It's slow even on a local Docker network, suggesting that the code ought to be reviewed. Locally, setting `?reduce_geometry=false` reduces the server response time insignificantly (**_with a smaller dataset than that tested on DEV_**), from 1.93s to 1.64s.

### Trying local modifications:

- Added index to local db using `CREATE INDEX idx_place_geom_geom_gist ON place_geom USING GIST (geom);` (already present on DEV), but this made little difference.
- TILER_URL response time is negligible (~.03s)
- Changed to `select_related` from `prefetch_related`, but this breaks the code.

Response is even worse for collections, and especially place collections where families have to be determined.

**Comment by docuracy** [comment]
Optimisation of `datasets.utils.fetch_mapdata_ds` still results in very long processing time, and so I have now implemented Django's `FileBasedCache` for this function and for `collection.views.fetch_mapdata_coll`. Server response times are reduced drastically after the first visit to a page:

Chinese Counties: 5.53s -> 29.53ms
Chinese Towns: 55.57s -> 515ms
Pokémon: 352ms -> 8.87ms

I've set the cache system (in `settings.py`) with no expiry, meaning that the functions will need to be called with `?refresh_cache=true` whenever the dataset or collection is updated, or whenever it has a tileset generated or removed. Calling them on first publication would of course remove the potentially-long wait on a first visit to the associated page.

Apart from evident performance gains, this will reduce the need to upgrade server processing and memory capacity. The disk capacity cost of maintaining a set of cache files (just one for each Dataset/Collection) is probably slight, but my security clearance level does not permit me to access the (new) `/cache` directory to assess the size of the generated files.

Modifications are currently running on DEV in a hotfix branch. I'll copy them to my local branch and make a PR in due course if you're satisfied with them, @kgeographer.

**Comment by docuracy** [comment]
After much trial-and-error hot-fixing, I've identified the source of the sometimes > 60-second delay in loading Place Detail pages, on the Dataset model:

```
  @property
  def minmax(self):
    # TODO: temporal is sparse, sometimes [None, None]
    timespans = [p.minmax for p in self.places.all() \
                 if p.minmax and len(p.minmax) == 2 and p.minmax[0]]
    # print('ds; timespans as model property', self.label, timespans)
    earliest = min([t[0] for t in timespans]) if len(timespans) > 0 else None
    latest = max([t[1] for t in timespans]) if len(timespans) > 0 else None
    minmax = [earliest, latest] if earliest and latest else None
    return minmax

```

I've replaced it with this, which deals with the #TODO and is _much_ more efficient:

```
  @property
  def minmax(self):
    minmax_values = Place.objects.filter(dataset=self).aggregate(
        # This ignores `None` values, effectively handling temporal sparsity [None, None]
        earliest=Min('minmax__0'),
        latest=Max('minmax__1')
    )
    earliest = minmax_values['earliest']
    latest = minmax_values['latest']
    return [earliest, latest] if earliest and latest else None
```

Server response time for https://dev.whgazetteer.org/places/5907944/detail is reduced from 62 seconds to 1.05 seconds.

**Comment by docuracy** [comment]
Search from Home Page is currently broken, because after a successful search and redirection to the Search Page, `search.js` seems not to be loading results from the `last_search` in LocalStorage.

**Comment by docuracy** [comment]
Moving ANNOUNCEMENTS block to the bottom of the Home Page has resulted in misalignment of the bottom of the columns.

**Comment by docuracy** [comment]
FYI, @kgeographer, some of the trace data (well, one trace in one collection that I know of, number 76) has a zero day number in a yearstring (which was breaking `fetch_mapdata_coll`). I've modified `collection.views.year_from_string` to handle it, but you might want to parse incoming data to eliminate things like that.

**Comment by docuracy** [comment]
Bug found when examining test dataset locally: even if this is just a test data problem, code should be made more robust to handle this eventuality.

http://localhost:8001/api/place/25195/

```
Traceback (most recent call last):
  File "/py/lib/python3.10/site-packages/django/core/handlers/exception.py", line 56, in inner
    response = get_response(request)
  File "/py/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/py/lib/python3.10/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/py/lib/python3.10/site-packages/django/views/generic/base.py", line 142, in dispatch
    return handler(request, *args, **kwargs)
  File "/app/api/views.py", line 1197, in get
    "types": add_urls(sort_unique([type for place in serialized_places for type in place["types"]], 'label')),
  File "/app/api/views.py", line 1130, in sort_unique
    item_unique = item[key] if key else json.dumps(item)

Exception Type: KeyError at /api/place/25195/
Exception Value: 'label'
```
